src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,144 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
+@SuppressWarnings("unchecked")
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private Object[] elementData;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+    }
+
+    public void rangeCheck() {
+        if (size == elementData.length) {
+            grow();
+        }
+    }
+
+    public void indexCheck(int index) {
+        if (index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+
+        if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    private Object[] grow() {
+        int newCapacity = elementData.length * 3 / 2 + 1; // Новий розмір (збільшення на 50%)
+        elementData = Arrays.copyOf(elementData, newCapacity); // Копіюємо елементи
+        return elementData;
+    }
+
+    public void checkIfElementExists(T element) {
+        boolean found = false;
+
+        // Перебираємо елементи масиву
+        for (int i = 0; i < size; i++) {
+            if ((element == null && elementData[i] == null)
+                    || (element != null && element.equals(elementData[i]))) {
+                found = true; // Елемент знайдено
+                break;
+            }
+        }
+
+        // Якщо елемент не знайдено, кидаємо виняток
+        if (!found) {
+            throw new NoSuchElementException("Element not found in the array.");
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        rangeCheck();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bound");
```

Comment:
The custom exception `ArrayListIndexOutOfBoundsException` is not defined. You need to create this exception class to handle invalid index cases as per the task requirements.